access/jira/app.go

```diff
@@ -59,320 +95,597 @@ func (a *App) WaitReady(ctx context.Context) (bool, error) {
-			log.Debugf("%v", trace.DebugReport(err))
-			return err
+		if err := a.onDeletedRequest(ctx, reqID); err != nil {
+			log.WithError(err).Errorf("Failed to process deleted request")
+			return trace.Wrap(err)
 		}
 		return nil
 	default:
 		return trace.BadParameter("unexpected event operation %s", op)
 	}
 }
 
-// onJIRAWebhook processes JIRA webhook and updates the status of an issue
-func (a *App) onJIRAWebhook(ctx context.Context, webhook Webhook) error {
-	log := logger.Get(ctx)
-
-	if webhook.WebhookEvent != "jira:issue_updated" || webhook.IssueEventTypeName != "issue_generic" {
+// onJiraWebhook processes Jira webhook and updates the status of an issue
+func (a *App) onJiraWebhook(ctx context.Context, webhook Webhook) error {
+	webhookEvent := webhook.WebhookEvent
+	issueEventTypeName := webhook.IssueEventTypeName
+	if webhookEvent != "jira:issue_updated" || issueEventTypeName != "issue_generic" {
 		return nil
 	}
 
+	ctx, log := logger.WithFields(ctx, logger.Fields{
+		"jira_issue_id": webhook.Issue.ID,
+	})
+	log.Debugf("Processing incoming webhook event %q with type %q", webhookEvent, issueEventTypeName)
+
 	if webhook.Issue == nil {
 		return trace.Errorf("got webhook without issue info")
 	}
 
-	issue, err := a.bot.GetIssue(ctx, webhook.Issue.ID)
+	issue, err := a.jira.GetIssue(ctx, webhook.Issue.ID)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	ctx, log = logger.WithFields(ctx, logger.Fields{
+		"jira_issue_id":  issue.ID,
+		"jira_issue_key": issue.Key,
+	})
 
 	statusName := strings.ToLower(issue.Fields.Status.Name)
-	if statusName == "pending" {
-		log.Debug("Issue is pending, ignoring it")
+	switch {
+	case statusName == "pending":
+		log.Debug("Issue has pending status, ignoring it")
 		return nil
-	} else if statusName == "expired" {
-		log.Debug("Issue is expired, ignoring it")
+	case statusName == "expired":
+		log.Debug("Issue has expired status, ignoring it")
 		return nil
-	} else if statusName != "approved" && statusName != "denied" {
-		return trace.BadParameter("unknown JIRA status %q", statusName)
+	case statusName != "approved" && statusName != "denied":
+		return trace.BadParameter("unknown Jira status %q", statusName)
 	}
 
-	reqID, err := issue.GetRequestID()
+	reqID, err := GetRequestID(issue)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	if reqID == "" {
+		log.Debugf("Missing %q issue property", RequestIDPropertyKey)
+		return nil
+	}
+
 	ctx, log = logger.WithField(ctx, "request_id", reqID)
 
-	req, err := a.accessClient.GetRequest(ctx, reqID)
+	reqs, err := a.apiClient.GetAccessRequests(ctx, types.AccessRequestFilter{ID: reqID})
 	if err != nil {
-		if trace.IsNotFound(err) {
-			log.WithError(err).Warning("Cannot process expired request")
-			return nil
-		}
 		return trace.Wrap(err)
 	}
 
-	if req.State != access.StatePending {
-		log.WithField("request_state", req.State).Warningf("Cannot process not pending request")
-		return nil
+	var req types.AccessRequest
+	if len(reqs) > 0 {
```

Comment:
Why do all the plugin data validation below if this condition could cause it to just fail (see line 341) anyway?

If that validation needs to happen regardless of whether `req == nil`, move this condition down closer to where the consequences happen. Alternatively, if all the validation can be avoided, bring the early exit up here.

Reply:
On the line 341 it's not a fail, it calls `resolveIssue`:

```go
	if req == nil {
		return trace.Wrap(a.resolveIssue(ctx, reqID, Resolution{Tag: ResolvedExpired}))
	}
```